Fix running LSP4iJ continuous integration builds with existing LTI releases

vaisakhkannan opened the pull request:
Fixes #815 

A part of the fix for - [#868](https://github.com/OpenLiberty/liberty-tools-intellij/issues/868)
Implementation of running LSP4iJ continuous integration builds with existing LTI releases

Currently, the fix is running against the branch `lsp4ij-market-0.0.2-integration`. We can add the `main` and other existing `LTI release tags`, but those will fail because of the `microshed:lsp4ij` dependency. Once we have an LTI release that uses the `redhat:lsp4ij` plugin, we will be able to easily add the version in the `tag` specified in the code.

vaisakhkannan commented:
Reference link for Cron job action : https://github.com/vaisakhkannan/liberty-tools-intellij/actions/runs/9856244732
link for default build : https://github.com/vaisakhkannan/liberty-tools-intellij/actions/runs/9856244722

aparnamichael commented on .github/workflows/build.yaml:
```diff
@@ -13,6 +13,17 @@ on:
+        type: string
+        required: true
+        default: main
+      # Specifying the default value as 'Liberty-tools-intellij' because we require the forked repository path for a successful build.
```
"because we require the forked repository path for a successful build." what does this comment mean?

vaisakhkannan replied:
I removed this comment from here , I already mentioned the need of this parameter in cronjob.yaml file line no `84`

aparnamichael commented on .github/workflows/cronJob.yaml:
```diff
@@ -74,16 +74,28 @@ jobs:
       fail-fast: false
       matrix:
+        # Existing LTI release tags can be added to obtain build results.
+        tag: [ 'lsp4ij-market-0.0.2-integration' ] # Can specify tags or branches such as '24.0.6' and 'main'
```
'tag' array is for specifying the LTI release tags. Can we run LTI builds on main branch even if 'tag' array is empty? 
rather than specifying 'main' inside the 'tag' array. Is there any way for that?

vaisakhkannan replied:
Yes, if 'tag' array is empty , the default value `main `will run.

aparnamichael commented on .github/workflows/build.yaml:
```diff
@@ -13,6 +13,17 @@ on:
+        required: true
+        default: main
+      # Specifying the default value as 'Liberty-tools-intellij' because we require the forked repository path for a successful build.
+      specifyRepo:
```
@vaisakhkannan This issue is for running LSP4IJ continuous integration builds with existing LTI releases. LTI releases/tags will be always in the openliberty repo. I am not sure this parameter will need once we merge these changes to main branch.

vaisakhkannan replied:
@aparnamichael , Yes, we don't need this `specifyRepo` once we integrate these changes into the `main` branch. I added this to ensure our tests work fine on both the `feature branch` and the `main` branch. Once this is merged into the main branch, the code will work as well. However, specifying this parameter is not necessary. Yes, it adds additional code. If it's not required, I can remove the specified parameter. For testing purposes, Anusree should add the code to her fork.

vaisakhkannan commented:
@anusreelakshmi934 , Link for the build --> https://github.com/vaisakhkannan/liberty-tools-intellij/actions/runs/9934992707

NOTE: With this latest change (based on the Kochi team's suggestion) Cron Job build will fail on fork cron job builds because `lsp4ij-market-0.0.2-integration` branch is not available in fork repo, because I removed the `SpecifyRepo` parameter, which is initially  used to specify the repository for the step named 'Checkout liberty-tools-intellij'. Since we are running on the main branch, it will default to `OpenLiberty/liberty-tools-intellij`. For testing purposes on a forked repository, the fork repository owner must Specify the repository as OpenLiberty/liberty-tools-intellij in build.yaml file as shown below. There won't be no issues with this code running on` main` branch.

`repository: OpenLiberty/Liberty-tools-intellij`

 Link For cron job when I added repository as `OpenLiberty/liberty-tools-intellij`  --> https://github.com/vaisakhkannan/liberty-tools-intellij/actions/runs/9934517308

anusreelakshmi934 commented:
> Link For cron job when I added repository as `OpenLiberty/liberty-tools-intellij` --> https://github.com/vaisakhkannan/liberty-tools-intellij/actions/runs/9934517308

Shouldn't the job also run against LTI main?

vaisakhkannan commented:
> > Link For cron job when I added repository as `OpenLiberty/liberty-tools-intellij` --> https://github.com/vaisakhkannan/liberty-tools-intellij/actions/runs/9934517308
> 
> Shouldn't the job also run against LTI main?

@anusreelakshmi934 , If we need to run on the main branch as well, we can add 'main' in the matrix tag.
like this
`tag: [ 'lsp4ij-market-0.0.2-integration', 'main' ]`

aparnamichael commented on .github/workflows/cronJob.yaml:
```diff
@@ -74,16 +74,25 @@ jobs:
       fail-fast: false
       matrix:
+        # Existing LTI release tags and branches can be added to obtain build results.If the tag array is empty, it will default to 'main'. However, if there is at least one tag or branch in the tag array and you need to run on 'main' as well, make sure to add 'main' to the array.
+        tag: [ 'lsp4ij-market-0.0.2-integration' ] # Can specify tags or branches such as '24.0.6' and 'main'
```
@TrevCraw While merging this change to main, we have to update the value in the tag. So, are you planning to keep the issue open till we merge these changes to main?

vaisakhkannan replied:
@aparnamichael , @TrevCraw suggested me to add a check box to this Epic: https://github.com/OpenLiberty/liberty-tools-intellij/issues/666

So I have added in this epic.

vaisakhkannan commented on .github/workflows/build.yaml:
```diff
@@ -96,7 +108,7 @@ jobs:
         uses: actions/upload-artifact@v4.3.4
         with:
-          name: liberty-tools-intellij-LTI-main-LSP4IJ-${{ env.LSP4IJ_BRANCH }}
+          name: liberty-tools-intellij-LTI-${{ env.REF_LTI_TAG }}-LSP4IJ-${{ env.LSP4IJ_BRANCH }}
```
As part of resolving the merge conflict, I found that this change is required. With this change, how will the artifact name look as shown in the screenshot below?
Cron Job build: Link : https://github.com/vaisakhkannan/liberty-tools-intellij/actions/runs/9988499251
<img width="838" alt="Screenshot 2024-07-18 at 2 02 21 PM" src="https://github.com/user-attachments/assets/2c3d444e-ea95-487c-b991-3514e835b59d">

Normal Build:  Link : https://github.com/vaisakhkannan/libe
<img width="1040" alt="Screenshot 2024-07-18 at 2 43 41 PM" src="https://github.com/user-attachments/assets/807bdc05-96c3-40ff-b0f1-56478db2c00d">
rty-tools-intellij/actions/runs/9988666800

The main point is, if this value `"${{ env.REF_LTI_TAG }}"` is empty ( see the image above), it means there is no value present. We can't simply put a default value like this: `LSP4IJ_BRANCH: ${{ inputs.lsp4ijBranch || 'latest' }}`, because `lsp4ijBranch `doesn’t require a default value (shown in the image below) in order to provide the 'latest' value.

<img width="675" alt="Screenshot 2024-07-18 at 2 49 06 PM" src="https://github.com/user-attachments/assets/69b905bd-2c6e-4ece-a2bc-96148859a24a">

Similarly, we can't provide a similar value because the `refLTITag` parameter requires a default value of 'main'.

TrevCraw replied:
A few questions/comments:

1. What does `latest` mean for the LSP4IJ branch? Does that mean using the `main` branch? If so, I think it would be better to use `main` in the artifact name.
2. It looks like the default value for `refLTITag` is `main`. How does this variable end up being empty?
3. If `refLTITag` is empty, how does the checkout part of the workflow work? https://github.com/OpenLiberty/liberty-tools-intellij/blob/76f216e3d88006553fdc9b9909c7d36dffe08f16/.github/workflows/build.yaml#L77-L81
4. If it is acceptable for `refLTITag` to be empty, is there a way to check if it is empty, and if so, assign it a value of `main` before using it in the artifact name or other places in the workflow?

vaisakhkannan replied:
@TrevCraw ,
1.` latest` doesn't mean that it is using the `main` branch.`latest` points to `lsp4ijVersion` https://github.com/OpenLiberty/liberty-tools-intellij/blob/b34b7af4d6db48c74c4d0bd5b92531919081dc73/build.gradle#L15.For normal builds --`USE_LOCAL_PLUGIN: false` Anusree made this change to get the value in normal builds (push/PR --> will print 'latest' value). She simply put 'latest' [here](https://github.com/OpenLiberty/liberty-tools-intellij/blob/b34b7af4d6db48c74c4d0bd5b92531919081dc73/.github/workflows/build.yaml#L57) because there is no default value specified in the [workflow_call](https://github.com/OpenLiberty/liberty-tools-intellij/blob/b34b7af4d6db48c74c4d0bd5b92531919081dc73/.github/workflows/build.yaml#L14).
2.

TrevCraw replied:
Thank you for the updates @vaisakhkannan. Can you change the name to be "default" for LSP4IJ as well? "latest" will not always be accurate, since it will be whatever is defined in the `build.gradle`.

vaisakhkannan replied:
Yes, Thank you @TrevCraw , I did the change.

aparnamichael reviewed (dismissed):
Looks good!!

anusreelakshmi934 commented:
Can you share the cronjob link here @vaisakhkannan ?

vaisakhkannan commented:
> Can you share the cronjob link here @vaisakhkannan ?

https://github.com/vaisakhkannan/liberty-tools-intellij/actions/runs/10072224523
https://github.com/vaisakhkannan/liberty-tools-intellij/actions/runs/10072710102
https://github.com/vaisakhkannan/liberty-tools-intellij/actions/runs/10071238101

anusreelakshmi934 reviewed (dismissed):
Looks good to me

mrglavas commented on .github/workflows/cronJob.yaml:
```diff
@@ -94,21 +94,30 @@ jobs:
+      fail-fast: false
+      matrix:
+        # Existing LTI release tags and branches can be added to obtain build results.If the tag array is empty, it will default to 'main'. However, if there is at least one tag or branch in the tag array and you need to run on 'main' as well, make sure to add 'main' to the array.
+        tag: [ 'lsp4ij-market-0.0.2-integration' ]
```
I assume the development branch being specified in the tag array is just a placeholder until we deliver this to `main`. We should open another issue to track the follow-up work, assuming you haven't done that already.

vaisakhkannan replied:
@mrglavas  , Anusree created an issue: https://github.com/OpenLiberty/liberty-tools-intellij/issues/895, and she added the following work item in list

**Remove `lsp4ij-market-0.0.2-integration` from matrix and add existing `LTI Tags/Branches` before merge to `main`**

mrglavas reviewed (approved):
Approving based on seeing this in action through demos and resolution of discussions with the team that I took part in. Great work!
